### PR TITLE
cloud: Re-architect the cloud package

### PR DESCRIPTION
**Pull request by ejj**
Before this patch, the database machine table contained the set of
    desired machines as populated by the engine.  And the cloud package
    was responsible for implementing that specification. This had a number
    of problems including:

        * Blueprint changes were reflected in the database and `quilt ps`
        immediately, before they had time to be enacted.  This confused
        users.

        * It was unnecessarily complex -- especially the `engine` package
        that had a confusing relationship with the cloud package.

        * The cloud joins were unnecessarily complicated as they had to
        carefully blend the desired machines with information from what
        was actually running.

This patch re-architects the cloud package with a new approach.  It
    now populates the database machine table with the set of machines that
    are running in the cloud provider.  And it joins that machine table
    with the set of desired machines as retrieved from the database
    blueprint table.  As a result:

        * Machines have a new "Stopping" status that shows in `quilt ps`
        when machines are being terminated.

        * The machine BlueprintID is no longer used, in favor of the ID
        provided by the cloud provider.  This even applies to the cli
        where users can cross-reference the IDs we show them with what
        their clod provider knows about.

        * Numerous other code deletions, simplifications, and cleanups.

### CODE REVIEW

**Comment by kklin** [comment]
I'll review this after Eli

**Comment by ejj** [comment]
FYI, I deleted a patch from this series as I realized a future patch I'm working on renders it obsolete.

**Comment by kklin** [comment]
The secrets implementation blocks on this (I'm just working on top of this for now)

**Comment by kklin** [comment]
FYI @ejj I'm working on top of your branch, and got this error after running `quilt stop`:
```
INFO [Oct  8 10:49:09.624] db.Machine:
        Machine-3{Master, Amazon us-west-1 m3.medium preemptible, sir-gw48ct5k, PublicIP=54.153.118.168, PrivateIP=172.31.5.229, Disk=32GB, stopping}
        Machine-2{Worker, Amazon us-west-1 m3.medium preemptible, sir-a29ie6kk, PublicIP=54.183.218.203, PrivateIP=172.31.13.87, Disk=32GB, stopping}
ERROR [Oct  8 10:49:14.323] Failed to get minion config              error=rpc error: code = Unavailable desc = grpc: the connection is unavailable
ERROR [Oct  8 10:49:14.323] Failed to get minion config              error=rpc error: code = Unavailable desc = grpc: the connection is unavailable
```

Looks to me like it's the same problem as with installing credentials when stopping machines: the foreman still tries to connect to the cluster while the machines are stopping.

**Comment by aegamesi** [comment]
`quilt stop` seems to work fine for me (and `quilt show` properly shows the 'stopping' status), but I'm getting a similar but different error from Kevin in the daemon:

```
INFO [Oct  8 13:26:23.038] Updated machines.                        count=2 action=stop region=DigitalOcean-sfo1-default-namespace
INFO [Oct  8 13:26:23.594] db.Machine:
                                   
ERROR [Oct  8 13:26:31.970] Failed to get minion config              error=rpc error: code = DeadlineExceeded desc = context deadline exceeded
ERROR [Oct  8 13:26:31.970] Failed to get minion config              error=rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

**Comment by ejj** [comment]
I updated with a tweak that should reduce the probability of this happening quite a bit.  I think there's a broader reworking of the foreman that's necessary here, but I don't think this is the patch to do it.

**Comment by aegamesi** [comment]
One thing I noticed while I was testing this on the nginx blueprint:
I booted up 1 master and 2 workers, and then I changed the blueprint to only have 1 worker. I ran the new blueprint, and then, while the third VM was being stopped, `quilt show` suddenly stopped showing information about my container:

```
$ quilt show
MACHINE         ROLE      PROVIDER    REGION        SIZE        PUBLIC IP         STATUS
quilt-c9d782    Master    Google      us-east1-b    f1-micro    35.196.141.122    connected
quilt-932a7e    Worker    Google      us-east1-b    f1-micro    35.185.19.1       connected
quilt-70b1f5    Worker    Google      us-east1-b    f1-micro    35.185.36.184     stopping

unable to query containers: rpc error: code = Unknown desc = Unable to connect to the Quilt daemon at tcp://35.185.36.184:9000: context deadline exceeded. Is the quilt daemon running? If not, you can start it with `quilt daemon`.
```

The container is still accessible (and remains accessible the entire time), the information just isn't available in `quilt show`.

**Comment by kklin** [comment]
@aegamesi @ejj That bug would happen because we need to filter by stopping here https://github.com/quilt/quilt/blob/master/api/server/server.go#L131 (and because that code could be a bit less scary about errors connecting)

File: cloud/join_test.go [hunk 1]
```diff
@@ -1,254 +1,253 @@
...
+		bp.Blueprint.Machines = []blueprint.Machine{{
+			Provider: string(FakeAmazon),
+			Region:   testRegion,
+			Size:     "1",
+		}}
+		bp.Namespace = "ns"
+		view.Commit(bp)
+		return nil
 	})
+	_, err = joinImpl(*cld)
```
Comment (kklin) [inline cloud/join_test.go]: What is this testing? Shouldn't you be checking whether the correct machines were specified in `boot`?

File: cloud/join_test.go [hunk 2]
```diff
@@ -1,254 +1,253 @@
...
+	m2.FloatingIP = m1.FloatingIP
+	m2.CloudID = "5"
+	assert.Equal(t, 1, machineScore(m1, m2))
+
+	// Wrong ID, but an assigned role.
+	m1 = m
+	m1.CloudID = "5"
+	assert.Equal(t, 2, machineScore(m, m2))
+
+	// Wrong IP, but no Role.
```
Comment (kklin) [inline cloud/join_test.go]: Should this be "Wrong ID, and no Role"?

File: cloud/credentials.go [hunk 3]
```diff
@@ -35,7 +35,9 @@ var credentialsCounter = counter.New("Cloud Credentials")
 func SyncCredentials(conn db.Conn, sshKey ssh.Signer, ca rsa.KeyPair) {
```
Comment (kklin) [inline cloud/credentials.go]: Should this be in its own commit? Shouldn't it be in the re-architecture commit because it fixes a bug introduced there?

File: cli/command/debug.go [hunk 4]
```diff
@@ -320,10 +320,10 @@ func containersToTargets(containers []db.Container, ips map[string]string) []log
 		}
 
 		t := logTarget{
-			ip:          ip,
-			dir:         containerDir,
-			blueprintID: c.BlueprintID,
-			cmds:        nil,
+			ip:      ip,
+			dir:     containerDir,
+			cloudID: c.BlueprintID,
```
[the remark this answers is not in the document]
Comment (ejj) [reply cli/command/debug.go]: Ah yeas that's much better ...

File: cloud/cloud.go [hunk 5]
```diff
@@ -126,43 +127,35 @@ func (cld cloud) run(stop <-chan struct{}) {
...
-			// running cloud machines that still need to communicate.
-			cld.syncACLs(jr.acls)
-			return
-		}
+	 * example) that should be reflected in the database, but won't be until
+	 * `runOnce()` is called a second time.  Lukily, these situations are nearly
+	 * always associated with machine status that cause a database trigger which will
+	 * cause the caller to invoke `runOnce()` again. */
+	jr, err := cloudJoin(cld)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (ejj) [reply cloud/cloud.go]: cloudJoin logs.

File: cloud/credentials.go (hunk 3 above)
Comment (ejj) [reply cloud/credentials.go]: Yeah now that I think about it, you're right.  I'll fold it in.

File: cloud/join.go [hunk 6]
```diff
@@ -42,52 +53,203 @@ func (cld cloud) join() (joinResult, error) {
...
+
+	for _, pair := range pairs {
+		dbm := pair.L.(db.Machine)
+		cm := pair.R.(db.Machine)
+
+		// Providers don't know about some fields, so we don't overwrite them.
+		cm.ID = dbm.ID
+		cm.Status = dbm.Status
+		cm.SSHKeys = dbm.SSHKeys
+		dbm = cm
```
[the remark this answers is not in the document]
Comment (ejj) [reply cloud/join.go]: For some odd reason I didn't think of it.

File: cloud/join.go [hunk 7]
```diff
@@ -42,52 +53,203 @@ func (cld cloud) join() (joinResult, error) {
...
+		view.Commit(dbm)
+	}
+}
+
+func (cld cloud) planUpdates(view db.Database) (boot, stop, updateIPs []db.Machine) {
+	bp, _ := view.GetBlueprint()
```
[the remark this answers is not in the document]
Comment (ejj) [reply cloud/join.go]: Yeah.  We already got the blueprint successfully once in this transaction.  I changed it so we check the error and panic.

File: cloud/join.go [hunk 8]
```diff
@@ -42,52 +53,203 @@ func (cld cloud) join() (joinResult, error) {
...
+		bpm := bpmi.(db.Machine)
+		dbm := view.InsertMachine()
+		bpm.ID = dbm.ID
+		bpm.Status = db.Booting
+		view.Commit(bpm)
+
+		boot = append(boot, bpm)
+	}
+
+	return
```
[the remark this answers is not in the document]
Comment (ejj) [reply cloud/join.go]: Huh.  I dunno.  I like it in this case, because it's impossible to get the order of the named arguments wrong.  Can't think of a reason against it.

[the remark this answers is not in the document]
Comment (kklin) [reply cloud/join.go]: SGTM, was just curious
